carcara/src/checker/rules/mod.rs

```diff
@@ -193,12 +195,42 @@ fn run_tests(test_name: &str, definitions: &str, cases: &[(&str, bool)]) {
         }));
 
         let mut checker = checker::ProofChecker::new(&mut pool, checker::Config::new());
-        let got = checker.check(&problem, &proof).is_ok();
-        assert_eq!(
-            *expected, got,
-            "test case \"{}\" index {} failed",
-            test_name, i
-        );
+        let check_result = checker.check(&problem, &proof);
+
+        // Extract error message, if any
+        let error_message = match &check_result {
+            Ok(_) => String::new(),
+            Err(crate::Error::Checker { inner, rule, step }) => match inner {
+                CheckerError::Explanation(s) => s.clone(),
+                _ => format!("[{rule} at {step}] {:?}", inner),
+            },
+            Err(crate::Error::DoesNotReachEmptyClause) => {
+                "Proof does not reach empty clause".into()
+            }
+            Err(e @ crate::Error::Parser(_, _)) => format!("Parser error: {}", e),
+            Err(e) => format!("Unexpected error: {:?}", e),
+        };
```

Comment:
I'm curious, why the different behaviour for the `Explanation` error variant? Also, I think it would be simpler to just format the `crate::Error` directly, instead of deconstructing it:
```suggestion
        let error_message = match &check_result {
            Ok(_) => String::new(),
            Err(e) => format!("{}", e),
        };
```
(are you doing it this way to get a more concise error message?)

Reply:
Ah, this "Explanation" error type is something Haniel showed me from his branch, that allows to express one of a kind error messages with an explanation. How could I do this in a function that may return a `RuleResult` ?

Reply:
Oh, I see. I think the existence of this error type makes sense then. Still, for the snippet I commented on, I think we can just format the error directly instead of matching on it.

Reply:
Ok, how would you do it with the `Explanation` variant?
Do you prefer that I commit this change?

Reply:
Yes. This suggestion just simplifies the code a bit, it will still format the `Explanation` errors correctly. More precisely, when you define the `Explanation` variant you used `#[error("{0}")]`, so if you print the `crate::Error` directly, it will print the `CheckError`, which will automatically print the contents of the `Explanation` error.